app/src/main/java/fr/free/nrw/commons/settings/SettingsFragment.java

```diff
@@ -144,28 +145,26 @@ private void prepareTheme() {
      * to remember later
      */
     private void prepareLanguages() {
-        List<String> languageNamesList = new ArrayList<>();
-        List<String> languageCodesList = new ArrayList<>();
-        List<Language> languages = getLanguagesSupportedByDevice();
-
-        for(Language language: languages) {
-            // Go through all languages and add them to lists
-            if(!languageCodesList.contains(language.getLocale().getLanguage())) {
-                // This if prevents us from adding same language twice
-                languageNamesList.add(language.getLocale().getDisplayName());
-                languageCodesList.add(language.getLocale().getLanguage());
-            }
+        List<String> languageNamesList;
+        List<String> languageCodesList;
+        AppLanguageLookUpTable appLanguageLookUpTable = new AppLanguageLookUpTable(getContext());
+        languageNamesList = appLanguageLookUpTable.getLocalizedNames();
+        languageCodesList = appLanguageLookUpTable.getCodes();
+        List<String> languageNameWithCodeList = new ArrayList<>();
+
+        for (int i = 0; i < languageNamesList.size(); i++) {
+            languageNameWithCodeList.add(languageNamesList.get(i) + "[" + languageCodesList.get(i) + "]");
```

Comment:
How does the upload activity generate this list?
Can we not factorize? If not, why?

Reply:
To generate the language list there is separate class AppLanguageLookUpTable.java 
Upload activity generate language list using  the getLocalizedName() and getCodes() function that is present in AppLanguageLookUpTable .java class 
I am also use this two function to generate the language list in setting activity

Reply:
`Upload activity generate language list`: Would you mind giving us the class name and line number of that? Thanks!

Reply:
`Upload activity generate language list:`class name :SpinnerLanguagesAdapter.java and line number :30

Reply:
The settings activity should reuse that `SpinnerLanguagesAdapter` (`SpinnerLanguagesAdapter` might require a few changes to be made reusable).

You can either:
- do it in this pull request,
- or create a new issue about that (and in that case I will merge this pull request because it still fixes the bug).

Reply:
SpinnerLanguagesAdapter is not reusable in setting activity becouse setting activity using  PreferenceScreen or preference .xml 
we do not implement  AppCompatSpinner in prefernce.xml
In setting activity ListPreference is use to show language list and we do not use AppCompatSpinner  in the place of ListPreference

Reply:
if you  have any idea to implement it   please tell me

Reply:
Oh you are right. It is fine like this, then. Thanks! :-)